Name: Dev Patel 

Course: DS4400 Data Mining and Machine Learning 1

Prof: Silvio Amir

University: Northeastern University 

## **Problem 5.1**

For each feature, write code to compute the average value, the min and max values, as well as its variance.
    
    Which features have the lowest and the highest average? Include the feature name and their average values for the features with the
    lowest and highest average.
    
    Which features have the lowest and the highest variance? Include the feature name and their variance values for the features with the
    lowest and highest variance.

In [9]:
import pandas as pd

# Load the entire dataset
df = pd.read_csv('kc_house_data.csv')

# Drop columns to ignore: id, date (non-features), and zipcode (categorical)
df = df.drop(['id', 'date', 'zipcode'], axis=1)

# Separate response variable and features
y = df['price']  # Response (dependent variable)
X = df.drop('price', axis=1)  # Features (independent variables)

In [10]:
# Compute statistics for each feature (excluding price - the response variable)
stats = pd.DataFrame({
    'Average': X.mean(),
    'Min': X.min(),
    'Max': X.max(),
    'Variance': X.var()
})

# Display the statistics table
stats

Average        Min           Max      Variance
bedrooms           3.370842     0.0000  3.300000e+01  8.650150e-01
bathrooms          2.114757     0.0000  8.000000e+00  5.931513e-01
sqft_living     2079.899736   290.0000  1.354000e+04  8.435337e+05
sqft_lot       15106.967566   520.0000  1.651359e+06  1.715659e+09
floors             1.494309     1.0000  3.500000e+00  2.915880e-01
waterfront         0.007542     0.0000  1.000000e+00  7.485226e-03
view               0.234303     0.0000  4.000000e+00  5.872426e-01
condition          3.409430     1.0000  5.000000e+00  4.234665e-01
grade              7.656873     1.0000  1.300000e+01  1.381703e+00
sqft_above      1788.390691   290.0000  9.410000e+03  6.857347e+05
sqft_basement    291.509045     0.0000  4.820000e+03  1.958727e+05
yr_built        1971.005136  1900.0000  2.015000e+03  8.627973e+02
yr_renovated      84.402258     0.0000  2.015000e+03  1.613462e+05
lat               47.560053    47.1559  4.777760e+01  1.919990e-02
long            -122.213896  -122.5190 -1.213150e+02  1.983262e-02
sqft_living15   1986.552492   399.0000  6.210000e+03  4.697612e+05
sqft_lot15     12768.455652   651.0000  8.712000e+05  7.455182e+08

In [11]:
# Find features with lowest and highest average
lowest_avg_feature = stats['Average'].idxmin()
highest_avg_feature = stats['Average'].idxmax()
lowest_avg_value = stats.loc[lowest_avg_feature, 'Average']
highest_avg_value = stats.loc[highest_avg_feature, 'Average']

print(f"the lowest average feature is {lowest_avg_feature} and the highest average feature is {highest_avg_feature}")
print(f"the lowest average value is {lowest_avg_value} and the highest average value is {highest_avg_value}")

the lowest average feature is long and the highest average feature is sqft_lot
the lowest average value is -122.21389640494147 and the highest average value is 15106.967565816869


## Features with Lowest and Highest Average

**Feature with LOWEST average:** `long` (longitude)  
**Average value:** -122.2139

**Feature with HIGHEST average:** `sqft_lot`  
**Average value:** 15,106.97

In [12]:
# Find features with lowest and highest variance
lowest_var_feature = stats['Variance'].idxmin()
highest_var_feature = stats['Variance'].idxmax()
lowest_var_value = stats.loc[lowest_var_feature, 'Variance']
highest_var_value = stats.loc[highest_var_feature, 'Variance']

print(f"The lowest feature variance is {lowest_var_feature} and the highest feature variance is {highest_var_feature}")
print(f"The lowest variance value is {lowest_var_value} and the highest variance value is {highest_var_value}")

The lowest feature variance is waterfront and the highest feature variance is sqft_lot
The lowest variance value is 0.007485225502689098 and the highest variance value is 1715658774.1754544


## Features with Lowest and Highest Variance

**Feature with LOWEST variance:** `waterfront`  
**Variance value:** 0.0075

**Feature with HIGHEST variance:** `sqft_lot`  
**Variance value:** 1,715,658,774.18 (approximately 1.72 billion)

## **Problem 5.2:** 

Compute the correlation coefficient of each feature with the response. Include a table with the correlation coefficient of each feature with the response. 

Which features are positively correlated (i.e., have positive correlation coefficient) with the response?

Which feature has the highest positive correlation with the response?

In [13]:
# Compute correlation coefficient of each feature with the response (price)
correlations = X.corrwith(y)

# Create a table sorted by correlation value
corr_table = pd.DataFrame({
    'Correlation with Price': correlations
}).sort_values(by='Correlation with Price', ascending=False)

corr_table

Correlation with Price
sqft_living                  0.702035
grade                        0.667434
sqft_above                   0.605567
sqft_living15                0.585379
bathrooms                    0.525138
view                         0.397293
sqft_basement                0.323816
bedrooms                     0.308350
lat                          0.307003
waterfront                   0.266369
floors                       0.256794
yr_renovated                 0.126434
sqft_lot                     0.089661
sqft_lot15                   0.082447
yr_built                     0.054012
condition                    0.036362
long                         0.021626

In [14]:
# Find features positively correlated with price
positive_corr = correlations[correlations > 0].sort_values(ascending=False)

# Feature with highest positive correlation
highest_corr_feature = positive_corr.idxmax()
highest_corr_value = positive_corr.max()

## Features Positively Correlated with Price

**All features** are positively correlated with the response (price):

| Feature | Correlation |
|---------|-------------|
| sqft_living | 0.7020 |
| grade | 0.6674 |
| sqft_above | 0.6056 |
| sqft_living15 | 0.5854 |
| bathrooms | 0.5251 |
| view | 0.3973 |
| sqft_basement | 0.3238 |
| bedrooms | 0.3083 |
| lat | 0.3070 |
| waterfront | 0.2664 |
| floors | 0.2568 |
| yr_renovated | 0.1264 |
| sqft_lot | 0.0897 |
| sqft_lot15 | 0.0824 |
| yr_built | 0.0540 |
| condition | 0.0364 |
| long | 0.0216 |

## Feature with Highest Positive Correlation

**Feature with HIGHEST positive correlation:** `sqft_living`  
**Correlation coefficient:** 0.7020

This makes intuitive sense - the living area square footage is the strongest predictor of house price.

## **Problem 5.3:** 

Were you able to find any features with a negative correlation coefficient with the response?

In [15]:
# Check for negatively correlated features
negative_corr = correlations[correlations < 0]
print(f"Number of features with negative correlation: {len(negative_corr)}")
if len(negative_corr) > 0:
    print("Negatively correlated features:")
    print(negative_corr.sort_values())
else:
    print("No features have a negative correlation with price.")

Number of features with negative correlation: 0
No features have a negative correlation with price.



**No**, there are no features with a negative correlation coefficient with the response (price).

All 17 features in the dataset have **positive correlations** with price, ranging from `long` (0.0216) to `sqft_living` (0.7020).

This suggests that for this dataset, increases in any of the feature values are associated with increases in house price (though some correlations are very weak, like `long` and `condition`).